feat(rust/rbac-registration): refactor cip509 extract key fns

no30bit opened the pull request:
# Description

Small refactoring of duplicate functions described in #288.

This PR introduces a new `Error` enum type. It replaces `anyhow::Error` return types where applicable in order to allow for higher order `ProblemReport` function calls.

Other than that, error messages are mostly maintained or made clearer, and the original logic remains.

## Related Issue(s)

Closes #288.

apskhem commented on rust/rbac-registration/src/cardano/cip509/utils/extract_key.rs:
```diff
@@ -48,47 +110,45 @@ pub fn x509_key(cert: &X509Certificate) -> anyhow::Result<VerifyingKey> {
+    let bytes = extended_public_key
+        .first_chunk::<PUBLIC_KEY_LENGTH>()
+        .ok_or_else(|| {
+            Error::Unexpected(anyhow!(
+                "Public key part length {PUBLIC_KEY_LENGTH} must be less \
+                than the extended length {EXTENDED_PUBLIC_KEY_LENGTH}"
+            ))
```
Why don't we use a more specific variant (or create one for this) rather than using `Unexpected` as we already know the exact cause (about pk length) on this? So this could get rid of the `Unexpected` variant from using `anyhow` as well.

no30bit replied:
It's just that a specific variant is a bit nonsensical.
That's a compile time checked thing.
I would use `Result::expect` if it was allowed in repo settings with a const assert.

no30bit replied:
Because the error boils down to one constant being less than the other against the common sense

stevenj replied:
Rather than a new error, either 
* just return `InvalidExtendedPublicKeyLength` there is no point in distinguishing the errors.  Because As you say, this second error must be handled, but is in fact impossible.
* Or In this case, I would also be 100% A-OK with an `.expect()` and allowing it with a directive to clippy for this line.  
Provided it has a comment explaining why `.expect()` is actually safe sp anyone working on this code understands why its safe, and we allowed it.

no30bit commented on rust/rbac-registration/src/cardano/cip509/utils/extract_key.rs:
```diff
@@ -1,19 +1,79 @@
+    #[error("Unsupported signature algorithm: {oid}")]
+    UnsupportedSignatureAlgo {
+        /// An OID of unsupported signature algorithm.
+        oid: String,
```
There's multiple `Oid` implementation in our use. 
Namely `oid_registry::Oid` and `spki::ObjectIdentifier`.
So, I assumed it would make sense to keep it a `String` as a "lingua-franca" intermediate representation.

stevenj replied:
`id_registry::Oid` is our preferred Oid implementation.

no30bit commented on rust/rbac-registration/src/cardano/cip509/utils/extract_key.rs:
```diff
@@ -22,24 +82,26 @@ pub struct SignatureAlgoError {
 ///
 /// Returns an error if the signature algorithm is not supported and
 /// the public key cannot be extracted.
-pub fn x509_key(cert: &X509Certificate) -> anyhow::Result<VerifyingKey> {
```
Great

no30bit replied:
Resolved

no30bit commented on rust/rbac-registration/src/cardano/cip509/utils/extract_key.rs:
```diff
@@ -22,24 +82,26 @@ pub struct SignatureAlgoError {
         .algorithm
         .oid
-        .to_string()
+        .to_string();
```
I didn't change the logic here.
This snippet transcodes between two 3rd-party `Oid` implementations.

no30bit replied:
Resolved (zerocost conversion instead of string roundtrip)

no30bit commented on rust/rbac-registration/src/cardano/cip509/utils/extract_key.rs:
```diff
@@ -22,24 +82,26 @@ pub struct SignatureAlgoError {
         .oid
-        .to_string()
+        .to_string();
+    let oid: Oid = oid_string
```
Ok, I checked, it could be done trivially 👍

no30bit replied:
Resolved (implemented conversion function)

no30bit commented:
Thx @stevenj, I will get back to you with the fixes shortly

stevenj commented on rust/rbac-registration/src/cardano/cip509/utils/extract_key.rs:
```diff
@@ -1,19 +1,79 @@
 
-/// Error type for unsupported signature algorithms.
+/// An extended public key length in bytes.
+const EXTENDED_PUBLIC_KEY_LENGTH: usize = 64;
```
Actually remove all reference to extended public keys here.  The certs can and must only contain regular ED25519 Public keys.

no30bit replied:
Resolved (no mention of extended keys anymore)

no30bit commented:
- I've removed all mentions of extended public keys.
In the only place I accept `subject_public_key` value I'm slicing the bytes by short public key length.
- I refactored the error type, removing redundant and unnecessary variants
- I've added a unit test for zerocost 3rd-party oid conversions

no30bit commented:
Can you look through the fixes @stevenj?

stanislav-tkach commented on rust/rbac-registration/src/cardano/cip509/utils/extract_key.rs:
```diff
@@ -48,47 +99,70 @@ pub fn x509_key(cert: &X509Certificate) -> anyhow::Result<VerifyingKey> {
+
+/// Creates [`VerifyingKey`] from the given public key.
+///
+/// Returns [`None`] if public key is incorrect in that context.
```
It seem the line about returning `None` can be removed.

no30bit replied:
Resolved

stanislav-tkach reviewed (approved):
Looking good to me, but please also wait for review from @stevenj and/or @Mr-Leshiy.

stevenj reviewed (dismissed):
LGTM

Mr-Leshiy commented on rust/rbac-registration/src/cardano/cip509/utils/extract_key.rs:
```diff
@@ -48,47 +99,68 @@ pub fn x509_key(cert: &X509Certificate) -> anyhow::Result<VerifyingKey> {
+    Oid::new(Cow::Borrowed(oid.as_bytes()))
+}
+
+/// Creates [`VerifyingKey`] from the given public key.
```
Add a comment that this function allow to construct  a `VerifyingKey` either from extended public key by just slicing `chain code` part, or from the common public key (32 bytes length).

no30bit replied:
Fixed

Mr-Leshiy commented on rust/rbac-registration/src/cardano/cip509/utils/extract_key.rs:
```diff
@@ -1,19 +1,75 @@
+    #[error("Invalid public key ({source})")]
+    PublicKeyIsNotEd25519 {
+        /// Underlying [`ed25519_dalek`] error.
+        #[source]
```
I rather use `from`, so it would allow to get rid of all the places with `map_err` and replace it with just `?`
```suggestion
        #[from]
```

Mr-Leshiy commented on rust/rbac-registration/src/cardano/cip509/utils/extract_key.rs:
```diff
@@ -48,47 +99,68 @@ pub fn x509_key(cert: &X509Certificate) -> anyhow::Result<VerifyingKey> {
+/// Creates [`VerifyingKey`] from the given public key.
+fn verifying_key(public_key: &[u8]) -> Result<VerifyingKey, Error> {
+    public_key
+        .first_chunk()
```
Add a TODO that `frist_chunk` should be replaced with `let bytes: [u8; 32] = public_key.try_into()?;` after we will support only common ed25119

no30bit replied:
Fixed

Mr-Leshiy commented on rust/rbac-registration/src/cardano/cip509/utils/extract_key.rs:
```diff
@@ -1,19 +1,75 @@
+        /// # Note
+        // Counting bits instead of bytes here, because of [`X509Certificate`] implementation of
+        // the public key storage.
+        bits: usize,
```
Nitpick.
I would replace `bits` with `bytes`.
It will avoid necessary conversion when this error is initialised and also it will be align with the `{PUBLIC_KEY_LENGTH} bytes` metric

Mr-Leshiy commented on rust/rbac-registration/src/cardano/cip509/utils/extract_key.rs:
```diff
@@ -22,24 +78,19 @@ pub struct SignatureAlgoError {
-        .context("Invalid subject_public_key value (has unused bits)")?;
-    verifying_key(extended_public_key).context("Unable to get verifying key from X509 certificate")
+        .subject_public_key;
+    let public_key_bytes = public_key.as_bytes().ok_or(Error::InvalidPublicKeyLength {
```
Nitpick.
Seems like not a totally correct reporting error variant/message. Should be something about not correctly aligned/has unused bits.

From the `as_bytes` docs

```
        /// Borrow the inner byte slice.
        ///
        /// Returns `None` if the number of unused bits is *not* equal to zero,
        /// i.e. if the `BIT STRING` is not octet aligned.
        ///
        /// Use [`BitString::raw_bytes`] to obtain access to the raw value
        /// regardless of the presence of unused bits.
        pub fn as_bytes(&self) -> Option<&[u8]> {
```

cong-or reviewed (approved):
LGTM

Mr-Leshiy reviewed (approved):
LGTM
